### PR TITLE
Us 2.1.11/password recovery

### PR DESCRIPTION
**Pull request by AlbertoPizzi**
Como usuario registrado quiero poder pedir modificar mi contraseña con un mecanismo seguro para poder recuperar el acceso a mi cuenta 

UATs:
- Se debe solicitar al usuario su email, en caso de que el email no exista igualmente el formulario se completa de manera normal.
- Si el email no está registrado el backend ignora el pedido de recuperación y no envía nada.
- Una vez completado el formulario se muestra el mensaje pidiendo al usuario que verifique su email.
- En el formulario de recupero de contraseña las restricciones son las mismas que en el alta de usuarios (requerido, mínimo 8, máximo 32, se debe verificar en el frontend pidiendo que se repita y las dos contraseñas provistas deben coincidir)

### CODE REVIEW

File: src/main/java/com/a2/backend/controller/UserController.java [hunk 1]
```diff
@@ -36,6 +36,13 @@ public UserController(UserService userService) {
         return ResponseEntity.status(HttpStatus.OK).body(userConfirmed);
     }
 
+    @PutMapping("/{email}")
+    public ResponseEntity<User> recoverPassword(
+            @PathVariable String email, @RequestBody PasswordRecoveryDTO passwordRecoveryDTO) {
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/controller/UserController.java]: why not `/recover` and let the email be part of the DTO?

File: src/main/java/com/a2/backend/entity/User.java [hunk 2]
```diff
@@ -31,5 +31,7 @@
 
     @JsonIgnore String confirmationToken;
 
+    String passwordRecoveryToken;
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/entity/User.java]: @JsonIgnore missing

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -52,6 +54,7 @@ public User createUser(UserCreateDTO userCreateDTO) {
                         .biography(userCreateDTO.getBiography())
                         .password(passwordEncoder.encode(userCreateDTO.getPassword()))
                         .confirmationToken(userCreateDTO.getConfirmationToken())
+                        .passwordRecoveryToken(randomStringUtils)
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/UserServiceImpl.java:59]: Token should be generated when the recovery process is started

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -107,4 +110,33 @@ public User confirmUser(String token, UUID userID) {
         user.setActive(true);
         return userRepository.save(user);
     }
+
+    @Override
+    public User recoverPassword(String email, PasswordRecoveryDTO passwordRecoveryDTO) {
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: This changes the password effectively, but when does the process start?

File: src/test/java/com/a2/backend/controller/UserControllerTest.java [hunk 5]
```diff
@@ -401,4 +401,60 @@ void Test010_GivenAValidTokenAndUserWhenConfirmingUserThenStatusOKisReturned() {
...
+                                userToBeUpdated.getId(),
+                                validConfirmationToken),
+                        HttpMethod.GET,
+                        null,
+                        User.class);
+        assertEquals(HttpStatus.OK, getResponse.getStatusCode());
+        passwordRecoveryDTO.setPasswordRecoveryToken(userToBeUpdated.getPasswordRecoveryToken());
+
+        User activatedUserToBeUpdated = getResponse.getBody();
+        System.out.println(userToBeUpdated.getPassword());
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/controller/UserControllerTest.java]: Why sout?

File: src/test/java/com/a2/backend/controller/UserControllerTest.java [hunk 6]
```diff
@@ -401,4 +401,60 @@ void Test010_GivenAValidTokenAndUserWhenConfirmingUserThenStatusOKisReturned() {
         assertEquals(
                 "There is an existing user with the nickname " + anotherNickname, errorMessage);
     }
+
+    @Test
+    void GivenAvalidUserWhenWantToRecoverPasswordThenReturnStatusOk() {
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/controller/UserControllerTest.java]: Overall the flow of the recovery process is wrong, it should be:
- GET recovery request to /user/recover
- At UserService generate a random recovery token and save it at the user model.
- Generated token is emailed to the user by a link to the frontend (localhost:3000/user/recover/{token})
- At the front end, user fills the form to request the password change
- POST request to /user/recover with a DTO containing: email, token and new password
- At UserService get user by email, verify the token matches and change password

**Review by agusayerza — CHANGES_REQUESTED** [review]
Incorrect flow

File: src/main/java/com/a2/backend/entity/User.java (hunk 2 above)
Comment (AlbertoPizzi) [reply src/main/java/com/a2/backend/entity/User.java]: For some reason the @JsonIgnore threw an error that I couldn't find the solution to.

File: src/test/java/com/a2/backend/controller/UserControllerTest.java (hunk 5 above)
Comment (AlbertoPizzi) [reply src/test/java/com/a2/backend/controller/UserControllerTest.java]: I forgot to delete that, I was getting null and wanted to check if it was the password's problem or what was it. Found the problem and fixed it, but forgot to delete the sout

File: src/test/java/com/a2/backend/controller/UserControllerTest.java (hunk 6 above)
Comment (AlbertoPizzi) [reply src/test/java/com/a2/backend/controller/UserControllerTest.java]: ok, so I'm not given the  DTO, the DTO is created at the userService impl?

Comment (agusayerza) [reply src/test/java/com/a2/backend/controller/UserControllerTest.java]: GET request should have a DTO containing the username of the user trying to be recovered. (maybe it should be a POST at /user/recover/request)
POST request should have the recover password DTO

File: src/main/java/com/a2/backend/model/PasswordRecoveryDTO.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+@AllArgsConstructor
+public class PasswordRecoveryDTO {
+
+    @Email(message = "Email must be valid")
+    String email;
+
+    @Size(min = 8, max = 32, message = "Password must be between 8 and 32 characters")
+    String newPassword;
+
+    @Size(min = 8, max = 32)
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/model/PasswordRecoveryDTO.java]: Acá me pongo gede por deporte no mas, pero ya sabes que el length tiene que ser 32 (ver comment de abajo)

File: src/main/java/com/a2/backend/service/impl/MailServiceImpl.java [hunk 8]
```diff
@@ -20,17 +20,24 @@ public MailServiceImpl(UserRepository userRepository) {
...
     }
 
     @Override
-    public void sendForgotPasswordMail(String mail) {}
+    public void sendForgotPasswordMail(User user) {
+        this.sendEmail(
+                user.getEmail(),
+                "Password Recovery",
+                "Hello in order to confirm your account press this link: "
+                        + "http://localhost:3000/user/recover/forgot-password?token="
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/MailServiceImpl.java:38]: Negociado con la persona de front, me imagino, no?

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java [hunk 9]
```diff
@@ -45,13 +51,15 @@ public User createUser(UserCreateDTO userCreateDTO) {
                     String.format(
                             "There is an existing user with the email %s",
                             userCreateDTO.getEmail()));
+        String randomStringUtils = RandomStringUtils.getAlphaNumericString(10);
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: 10 nada mas? Mandale 32

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java [hunk 10]
```diff
@@ -107,4 +115,41 @@ public User confirmUser(String token, UUID userID) {
         user.setActive(true);
         return userRepository.save(user);
     }
+
+    @Override
+    public User recoverPassword(PasswordRecoveryDTO passwordRecoveryDTO) {
+        val userOptional = userRepository.findByEmail(passwordRecoveryDTO.getEmail());
+        if (userOptional.isEmpty()) {
+            return null;
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: null?

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java [hunk 11]
```diff
@@ -107,4 +115,41 @@ public User confirmUser(String token, UUID userID) {
...
+        if (!userOptional
+                .get()
+                .getPasswordRecoveryToken()
+                .equals(passwordRecoveryDTO.getPasswordRecoveryToken())) {
+            throw new TokenConfirmationFailedException(
+                    String.format(
+                            "Invalid Token %s", passwordRecoveryDTO.getPasswordRecoveryToken()));
+        }
+        if (!userOptional.get().isActive()) {
+            return null;
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: null?

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java [hunk 12]
```diff
@@ -107,4 +115,41 @@ public User confirmUser(String token, UUID userID) {
...
+        }
+        if (passwordRecoveryDTO.getNewPassword().length() < 8
+                || passwordRecoveryDTO.getNewPassword().length() > 32) {
+            return null;
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: null?

Comment (agusayerza) [reply src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: Hacete solo una PasswordRecoveryFailedException y tirala con mismo mensaje

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java [hunk 13]
```diff
@@ -107,4 +115,41 @@ public User confirmUser(String token, UUID userID) {
...
+        }
+        val userToUpdatePassword = userOptional.get();
+
+        userToUpdatePassword.setPassword(
+                passwordEncoder.encode(passwordRecoveryDTO.getNewPassword()));
+        String randomStringUtils = RandomStringUtils.getAlphaNumericString(10);
```
Comment (agusayerza) [inline src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: 32

File: src/test/java/com/a2/backend/controller/UserControllerTest.java [hunk 14]
```diff
@@ -254,13 +263,15 @@ void Test010_GivenAValidTokenAndUserWhenConfirmingUserThenStatusOKisReturned() {
         val activatedUser = getResponse.getBody();
         assertNotNull(activatedUser);
         assertTrue(activatedUser.isActive());
+
+        System.out.println("This is a password: " + userToActivate.getPassword());
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/controller/UserControllerTest.java]: remove

File: src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java [hunk 15]
```diff
@@ -164,6 +174,7 @@ void Test007_GivenAValidValidTokenWhenConfirmAccountThenUserIsActiveEqualsTrue()
         User activatedUser = userService.confirmUser(confirmationToken1, user.getId());
 
         assertTrue(activatedUser.isActive());
+        System.out.println("this is the password: " + activatedUser.getPassword());
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java]: remove

File: src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java [hunk 16]
```diff
@@ -258,4 +269,64 @@ void Test011_GivenANonExistentUserWhenUpdatingThenExceptionIsThrown() {
...
+        User user = userService.createUser(userCreateDTO);
+        passwordRecoveryDTO.setPasswordRecoveryToken(user.getPasswordRecoveryToken());
+
+        User userToBeUpdated = userService.confirmUser("token001", user.getId());
+
+        User passwordUpdatedUser = userService.recoverPassword(passwordRecoveryDTO);
+
+        assertNotEquals(userToBeUpdated.getPassword(), passwordUpdatedUser.getPassword());
+
+        System.out.println(passwordUpdatedUser.getPassword());
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java:282]: assert?

File: src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java [hunk 17]
```diff
@@ -258,4 +269,64 @@ void Test011_GivenANonExistentUserWhenUpdatingThenExceptionIsThrown() {
...
+    }
+
+    @Test
+    void Test010_GivenAnInactiveUserWhenWantToRecoverPassWordThenReturnNull() {
+        User user = userService.createUser(userCreateDTO);
+        passwordRecoveryDTO.setPasswordRecoveryToken(user.getPasswordRecoveryToken());
+
+        User passwordUpdatedUser = userService.recoverPassword(passwordRecoveryDTO);
+
+        assertEquals(null, passwordUpdatedUser);
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java]: catch exception

File: src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java [hunk 18]
```diff
@@ -258,4 +269,64 @@ void Test011_GivenANonExistentUserWhenUpdatingThenExceptionIsThrown() {
...
+    }
+
+    @Test
+    void Test011_GivenAnInvalidEmailWhenWantToRecoverPasswordThenReturnNull() {
+        passwordRecoveryDTO.setEmail("adihajkd");
+        User passwordUpdatedUser = userService.recoverPassword(passwordRecoveryDTO);
+        assertEquals(null, passwordUpdatedUser);
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java]: idem

File: src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java [hunk 19]
```diff
@@ -258,4 +269,64 @@ void Test011_GivenANonExistentUserWhenUpdatingThenExceptionIsThrown() {
...
+        passwordRecoveryDTO.setNewPassword("test001");
+        User userToBeUpdated = userService.confirmUser("token001", user.getId());
+        passwordRecoveryDTO.setEmail(userToBeUpdated.getEmail());
+        User passwordUpdatedUser = userService.recoverPassword(passwordRecoveryDTO);
+
+        assertEquals(null, passwordUpdatedUser);
+    }
+
+    @Test
+    void Test014_GivenAnIntWhenWantARandomStringThenCreatesTheRandomString() {
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java]: Test without assert?

File: src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java [hunk 20]
```diff
@@ -258,4 +269,64 @@ void Test011_GivenANonExistentUserWhenUpdatingThenExceptionIsThrown() {
...
+        String randomStringTest = RandomStringUtils.getAlphaNumericString(10);
+        System.out.println(randomStringTest);
```
Comment (agusayerza) [inline src/test/java/com/a2/backend/service/impl/UserServiceImplTest.java]: remove

**Review by agusayerza — CHANGES_REQUESTED** [review]
Pequeños cambios, nada grave

File: src/main/java/com/a2/backend/model/PasswordRecoveryDTO.java (hunk 7 above)
Comment (AlbertoPizzi) [reply src/main/java/com/a2/backend/model/PasswordRecoveryDTO.java]: Cual comment de abajo ?

File: src/main/java/com/a2/backend/service/impl/UserServiceImpl.java (hunk 9 above)
Comment (agusayerza) [reply src/main/java/com/a2/backend/service/impl/UserServiceImpl.java]: @AlbertoPizzi

File: src/main/java/com/a2/backend/model/PasswordRecoveryDTO.java (hunk 7 above)
Comment (agusayerza) [reply src/main/java/com/a2/backend/model/PasswordRecoveryDTO.java]: Lo que digo es que hagas `min = 32, max = 32`

File: src/main/java/com/a2/backend/service/impl/MailServiceImpl.java (hunk 8 above)
Comment (AlbertoPizzi) [reply src/main/java/com/a2/backend/service/impl/MailServiceImpl.java:38]: Si lo hable con Andres y tengo entendido que esta bien el link

**Review by agusayerza — APPROVED** [review]

**Pull request merged**